feat(parameters): migrate AppConfig to new APIs due to API deprecation

leandrodamascena opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #1506 

## Summary

### Changes

Due to [Deprecation of the AppConfig API](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_GetConfiguration.html), we needed to change the SDK client to call the new APIs.

We had a long discussion in this issue #1506 before applying these changes.

### User experience

No changes are required to Lambda functions that use `Feature Flags and AppConfig Parameter utility`, but users must add `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions for this to work in their Lambdas.

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x ] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x ] I have performed a self-review of this change
* [x ] Changes have been tested
* [x ] Changes are documented
* [x ] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.


-----
[View rendered docs/upgrade.md](https://github.com/leandrodamascena/aws-lambda-powertools-python/blob/feat/new-appconfig-apis/docs/upgrade.md)

leandrodamascena committed:
feat(v2/appconfigdata): refactoring new api calls

leandrodamascena committed:
feat(v2/appconfigdata): refactoring new api calls

leandrodamascena commented on tests/e2e/parameters/conftest.py:
```diff
@@ -0,0 +1,19 @@
+import pytest
+
+from tests.e2e.parameters.infrastructure import ParametersStack
```
I don't know what's going on, but I couldn't add the other commits with more e2e tests. I'm fixing it locally and will add it soon.

leandrodamascena committed:
feat(v2/appconfigdata): adding more e2e tests

leandrodamascena commented on docs/upgrade.md:
```diff
@@ -154,3 +155,9 @@ Prior to this change, the Idempotency key was generated using only the caller fu
 
 Using qualified names prevents distinct functions with the same name to contend for the same Idempotency key.
+
+## Feature Flags and AppConfig Parameter utility
```
Hi @heitorlessa! I need help to improve the documentation. Maybe adding best practices using all parameters under the same configuration?

heitorlessa replied:
Sure. We should add an `Info` Admonition in the Parameters and Feature Flags docs.

ran-isenberg commented on aws_lambda_powertools/utilities/parameters/appconfig.py:
```diff
@@ -100,16 +100,23 @@ def _get(self, name: str, **sdk_options) -> str:
-        sdk_options["Application"] = self.application
-        sdk_options["Environment"] = self.environment
-        sdk_options["ClientId"] = CLIENT_ID
+        response = self.client.get_latest_configuration(ConfigurationToken=self.next_call)
```
arent you supposed to add here the sdk_options? let's say i want to config from a different region or some other config

leandrodamascena replied:
Thanks for pointing out this @ran-isenberg. Configuration like region, retries, credentials and others are passed to the constructor and not to the API calls, that is, if you check [this line](https://github.com/awslabs/aws-lambda-powertools-python/pull/1553/files#diff-5576360015200fe47fe1113f8fcee3d78e9fdb6059051e7663d220c7dadcf78bR82) you see any additional configuration is passed to the constructor.

I created this fictional code to make sure we are not missing anything
![image](https://user-images.githubusercontent.com/4295173/193814001-f9c698e9-f406-4d2a-a290-fd9517e20cd8.png)
![image](https://user-images.githubusercontent.com/4295173/193814746-4fa042e6-d07d-4c48-bf5e-4950285de7c8.png)
![image](https://user-images.githubusercontent.com/4295173/193814439-349203e7-8f9a-44a1-ba73-f767fc8327cb.png)

I think you mentioned a situation where a customer needs to pass [additional parameters to the SDK](https://awslabs.github.io/aws-lambda-powertools-python/latest/utilities/parameters/#passing-additional-sdk-arguments) and it makes sense, but I don't think that's the situation here. The method [get_latest_configuration](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/appconfigdata.html#AppConfigData.Client.get_latest_configuration) only supports the `ConfigurationToken` argument and if in the future it supports more than this argument, we need to review the impact of a possible change.

ran-isenberg replied:
yes, my bad! I looked at the wrong api doc. sorry about that.

leandrodamascena committed:
Merge branch 'v2' into feat/new-appconfig-apis

leandrodamascena committed:
Merge branch 'v2' into feat/new-appconfig-apis

leandrodamascena commented:
Giving feedback here. 
I need to add more information in the doc files, but we hope to review and merge this PR later this week.

leandrodamascena committed:
feat(v2/appconfigdata): fixing small things

leandrodamascena committed:
Merge branch 'v2' into feat/new-appconfig-apis

leandrodamascena committed:
feat(v2/appconfigdata): adding documentation and fix e2e tests

leandrodamascena committed:
feat(v2/appconfigdata): removing Any typing

heitorlessa commented:
Looking into this now [apologies for the delay, re:Invent deadlines are tighter this year]

heitorlessa commented on aws_lambda_powertools/utilities/parameters/appconfig.py:
```diff
@@ -100,16 +100,26 @@ def _get(self, name: str, **sdk_options) -> str:
             Name of the configuration
         sdk_options: dict, optional
-            Dictionary of options that will be passed to the client's get_configuration API call
+            Dictionary of options that will be passed to the client's start_configuration_session API call
```
```suggestion
            SDK options to propagate to `start_configuration_session` API call
```

Fixing older and suboptimal wording (when this was introduced)

heitorlessa commented on aws_lambda_powertools/utilities/parameters/appconfig.py:
```diff
@@ -91,6 +88,9 @@ def __init__(
         self.environment = environment
         self.current_version = ""
 
+        self.next_call = ""
```
```suggestion
        self._next_token = ""
```

Unless we're planning to expose the next token to customers, let's make it private.

heitorlessa commented on aws_lambda_powertools/utilities/parameters/appconfig.py:
```diff
@@ -145,7 +155,7 @@ def get_app_config(
         Maximum age of the cached value
     sdk_options: dict, optional
-        Dictionary of options that will be passed to the boto client get_configuration API call
+        Dictionary of options that will be passed to the boto client start_configuration_session API call
```
```suggestion
        SDK options to propagate to `start_configuration_session` API call
```

heitorlessa commented on docs/upgrade.md:
```diff
@@ -13,6 +13,7 @@ Changes at a glance:
 * The API for **event handler's `Response`** has minor changes to support multi value headers and cookies.
 * The **legacy SQS batch processor** was removed.
 * The **Idempotency key** format changed slightly, invalidating all the existing cached results.
+* The **Feature Flags and AppConfig Parameter utility** API calls have changed and you must add new IAM permissions to Lambda.
```
```suggestion
* The **Feature Flags and AppConfig Parameter utility** API calls have changed and you must update your IAM permissions.
```

heitorlessa commented on docs/upgrade.md:
```diff
@@ -154,3 +155,9 @@ Prior to this change, the Idempotency key was generated using only the caller fu
+
+The current API is [deprecated](https://docs.aws.amazon.com/appconfig/2019-10-09/APIReference/API_GetConfiguration.html) for new accounts and we had to change this.
+
+No changes are required to your code, but you must add `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions for this to work in your Lambdas.
```
```suggestion
AWS AppConfig deprecated the current API (GetConfiguration) - [more details here](https://github.com/awslabs/aws-lambda-powertools-python/issues/1506#issuecomment-1266645884).

You must update your IAM permissions to allow `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession`. There are no code changes required.
```

heitorlessa commented on docs/utilities/feature_flags.md:
```diff
@@ -27,6 +27,10 @@ If you want to learn more about feature flags, their variations and trade-offs,
 * [Feature Toggles (aka Feature Flags) - Pete Hodgson](https://martinfowler.com/articles/feature-toggles.html)
 * [AWS Lambda Feature Toggles Made Simple - Ran Isenberg](https://isenberg-ran.medium.com/aws-lambda-feature-toggles-made-simple-580b0c444233)
 * [Feature Flags Getting Started - CloudBees](https://www.cloudbees.com/blog/ultimate-feature-flag-guide)
+* [Best Practices for validating AWS AppConfig Feature Flags and Configuration Data](https://aws.amazon.com/pt/blogs/mt/best-practices-for-validating-aws-appconfig-feature-flags-and-configuration-data/)
```
```suggestion
```

Not applicable here. You also accidentally shared a `pt` domain

heitorlessa commented on docs/utilities/feature_flags.md:
```diff
@@ -27,6 +27,10 @@ If you want to learn more about feature flags, their variations and trade-offs,
+* [Best Practices for validating AWS AppConfig Feature Flags and Configuration Data](https://aws.amazon.com/pt/blogs/mt/best-practices-for-validating-aws-appconfig-feature-flags-and-configuration-data/)
+
+???+ note
+    Optimize your Lambda execution time by putting your feature settings in a single [Configuration](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html).
```
```suggestion
    AWS AppConfig requires two API calls to fetch configuration for the first time. You can improve latency  by consolidating your feature settings in a single [Configuration](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html).
```

In this case, a tip is to share why they should continue reading this note - Challenge first, recommended step afterward. If the challenge doesn't concern them, they can stop reading immediately. This optimizes for cognitive load.

heitorlessa commented on docs/utilities/feature_flags.md:
```diff
@@ -38,7 +42,7 @@ If you want to learn more about feature flags, their variations and trade-offs,
 ### IAM Permissions
 
-Your Lambda function must have `appconfig:GetConfiguration` IAM permission in order to fetch configuration from AWS AppConfig.
+Your Lambda function must have `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions in order to fetch configuration from AWS AppConfig.
```
```suggestion
Your Lambda function IAM Role must have `appconfig:GetLatestConfiguration` and `appconfig:StartConfigurationSession` IAM permissions before using this feature.
```

fixing older wording

heitorlessa commented on tests/functional/test_utilities_parameters.py:
```diff
@@ -1639,14 +1639,22 @@ def test_appconf_provider_get_configuration_json_content_type(mock_name, config)
-    response = {"Content": mock_value, "ConfigurationVersion": "1", "ContentType": "application/json"}
-    stubber.add_response("get_configuration", response)
+    response_start = {"InitialConfigurationToken": "initial_token"}
+    stubber.add_response("start_configuration_session", response_start)
```
```suggestion
    response_start_config_session = {"InitialConfigurationToken": "initial_token"}
    stubber.add_response("start_configuration_session", response_start_config_session)
```

easier to read which response belongs to which API

heitorlessa commented on tests/functional/test_utilities_parameters.py:
```diff
@@ -1639,14 +1639,22 @@ def test_appconf_provider_get_configuration_json_content_type(mock_name, config)
+        "NextPollConfigurationToken": "initial_token",
+        "ContentType": "application/json",
+    }
+    stubber.add_response("get_latest_configuration", response_latest)
```
```suggestion
    response_get_latest_config = {
        "Configuration": mock_value,
        "NextPollConfigurationToken": "initial_token",
        "ContentType": "application/json",
    }
    stubber.add_response("get_latest_configuration", response_get_latest_config)
```

easier to read which response belongs to which API

heitorlessa commented on tests/functional/test_utilities_parameters.py:
```diff
@@ -1670,14 +1678,22 @@ def test_appconf_provider_get_configuration_json_content_type_with_custom_client
-    response = {"Content": mock_value, "ConfigurationVersion": "1", "ContentType": "application/json"}
-    stubber.add_response("get_configuration", response)
+    response_start = {"InitialConfigurationToken": "initial_token"}
+    stubber.add_response("start_configuration_session", response_start)
```
```suggestion
    response_start_config_session = {"InitialConfigurationToken": "initial_token"}
    stubber.add_response("start_configuration_session", response_start_config_session)
```

heitorlessa commented on tests/functional/test_utilities_parameters.py:
```diff
@@ -1670,14 +1678,22 @@ def test_appconf_provider_get_configuration_json_content_type_with_custom_client
+        "NextPollConfigurationToken": "initial_token",
+        "ContentType": "application/json",
+    }
+    stubber.add_response("get_latest_configuration", response_latest)
```
```suggestion
    response_get_latest_config = {
        "Configuration": mock_value,
        "NextPollConfigurationToken": "initial_token",
        "ContentType": "application/json",
    }
    stubber.add_response("get_latest_configuration", response_get_latest_config)
```

heitorlessa commented on tests/functional/test_utilities_parameters.py:
```diff
@@ -1699,10 +1715,16 @@ def test_appconf_provider_get_configuration_no_transform(mock_name, config):
+        "NextPollConfigurationToken": "initial_token",
+        "ContentType": "application/json",
+    }
+    stubber.add_response("get_latest_configuration", response_latest)
```
```suggestion
    response_get_latest_config = {
        "Configuration": mock_value,
        "NextPollConfigurationToken": "initial_token",
        "ContentType": "application/json",
    }
    stubber.add_response("get_latest_configuration", response_get_latest_config)
```

heitorlessa commented on tests/functional/test_utilities_parameters.py:
```diff
@@ -1699,10 +1715,16 @@ def test_appconf_provider_get_configuration_no_transform(mock_name, config):
-    response = {"Content": mock_value, "ConfigurationVersion": "1", "ContentType": "application/json"}
-    stubber.add_response("get_configuration", response)
+    response_start = {"InitialConfigurationToken": "initial_token"}
+    stubber.add_response("start_configuration_session", response_start)
```
```suggestion
    response_start_config_session = {"InitialConfigurationToken": "initial_token"}
    stubber.add_response("start_configuration_session", response_start_config_session)
```

heitorlessa commented on tests/e2e/parameters/test_ssm.py:
```diff
@@ -0,0 +1,27 @@
+import json
+
+import pytest
+
+from tests.e2e.utils import data_fetcher
+
+
+@pytest.fixture
+def parameter_string_handler_fn_arn(infrastructure: dict) -> str:
+    return infrastructure.get("ParameterStringHandlerArn", "")
+
+
+@pytest.fixture
+def parameter_string_value(infrastructure: dict) -> str:
+    return infrastructure.get("ParameterStringValue", "")
+
+
+def test_simple_parameter_string(parameter_string_handler_fn_arn: str, parameter_string_value: str):
+    # GIVEN
+    expected_return = json.dumps({"value": parameter_string_value})
+
+    # WHEN
+    parameter_execution, _ = data_fetcher.get_lambda_response(lambda_arn=parameter_string_handler_fn_arn)
+    parameter_value = parameter_execution["Payload"].read().decode("utf-8")
+
+    # THEN
+    assert parameter_value == expected_return
```
could you split this into a separate PR? If we need to revert this AppConfig PR, we won't affect SSM E2E. It also helps us to have a better focus on SSM `with_decryption=True` too.

leandrodamascena replied:
I agree 100% @heitorlessa! I removed it and I will raise a new issue for this.

heitorlessa commented on tests/e2e/parameters/test_appconfig.py:
```diff
@@ -0,0 +1,36 @@
+def parameter_appconfig_freeform_value(infrastructure: dict) -> str:
+    return infrastructure.get("AppConfigConfigurationValue", "")
+
+
```
Should we have a test for FeatureFlags ConfigurationProfile type too?

I don't recall if we have full control on the content of that type in AppConfig (as in, add our expected schema). If we do, it'd be contribute to a more comprehensive coverage to detect this feature will work in both scenarios.

PS: The fetching will be exactly the same, so your Lambda function can be reused (remove `freeform` from the handlers name)

leandrodamascena replied:
@heitorlessa currently we don't support [AWS.AppConfig.FeatureFlags](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html#appconfig-type-reference-feature-flags) profile and I added a info in [documentation](https://github.com/awslabs/aws-lambda-powertools-python/pull/1553/files#diff-0e8fe1a0c18945a2f28b5a87df6c635fbbbfc7d87e44305a0978e41db488024bR8)

heitorlessa commented on tests/e2e/parameters/infrastructure.py:
```diff
@@ -0,0 +1,103 @@
+        cfn_deployment_strategy = appconfig.CfnDeploymentStrategy(
+            self.stack,
+            "appconfig-deployment-strategy",
+            deployment_duration_in_minutes=0,
+            final_bake_time_in_minutes=0,
```
Superb!!

heitorlessa commented on tests/e2e/parameters/infrastructure.py:
```diff
@@ -0,0 +1,103 @@
+            deployment_duration_in_minutes=0,
+            final_bake_time_in_minutes=0,
+            growth_factor=100,
+            name="deploymente2e",
```
I'm 98% certain this won't be a problem, but I wonder if AppConfig allows multiple Deployment Strategies with the same name in the event of parallel tests running (3.7-3.9).

If we do see any CI error, we can easily use our builder to generate random values.

heitorlessa commented on tests/e2e/parameters/infrastructure.py:
```diff
@@ -0,0 +1,103 @@
+
+    def _create_app_config(self, function: Function):
+
+        cfn_application = appconfig.CfnApplication(self.stack, id="appconfig-app", name="appe2e", description="appe2e")
```
Same as my other comment - do you think we can run into an issue with the name being hard coded?

e.g., parallel tests, tests that failed due to networking and left orphan resources, etc.

leandrodamascena replied:
@heitorlessa thanks for bringing all these insights! This is very important so that I don't make these mistakes again in the future. :heart:

heitorlessa commented on tests/e2e/parameters/infrastructure.py:
```diff
@@ -0,0 +1,103 @@
+
+    def _create_parameter_string(self, function: Function):
+        parameter = ssm.StringParameter(
+            self.stack, id="string_parameter", parameter_name="sample_string", string_value="Lambda Powertools"
```
We need to split SSM into a separate PR. When you do, we need to ensure the parameter name is unique otherwise it'll fail subsequent tests running in parallel

leandrodamascena replied:
I agree 100% @heitorlessa! I removed it and I will raise a new issue for this.

heitorlessa reviewed (changes requested):
Thanks a lot @leandrodamascena for working on this, truly!

I've made some wording suggestions, and some concerns on E2E tests:

* Separate SSM from this PR for a clean cut and revert if we need
* Some resources have names hardcoded, this might bite us back in parallel testing or network failures (e.g., SSM will fail subsequent stack if a param name already exists)

leandrodamascena committed:
feat(v2/appconfigdata): wording + variable name

leandrodamascena committed:
feat(v2/appconfigdata): wording + variable name

leandrodamascena committed:
feat(v2/appconfigdata): e2e changes

leandrodamascena committed:
feat(v2/appconfigdata): small fix documentation

leandrodamascena commented:
> Thanks a lot @leandrodamascena for working on this, truly!
> 
> I've made some wording suggestions, and some concerns on E2E tests:
> 
> * Separate SSM from this PR for a clean cut and revert if we need
> * Some resources have names hardcoded, this might bite us back in parallel testing or network failures (e.g., SSM will fail subsequent stack if a param name already exists)

Heiiiiii @heitorlessa! I've addressed all the feedback and I think we're ready to merge.

Thank you so much for reviewing this, every time you give me WONDERFUL feedback to think about UX and see beyond the code. :heart: 

![image](https://user-images.githubusercontent.com/4295173/195865829-6f9b4243-f798-440e-8f0e-ece0088f98ca.png)

heitorlessa commented on docs/utilities/feature_flags.md:
```diff
@@ -3,11 +3,12 @@ title: Feature flags
 
+???+ info
+    We currently only support AppConfig using [freeform configuration profile](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html#appconfig-creating-configuration-and-profile-free-form-configurations).
+    We are planning to add support for the [feature flag configuration profile](https://docs.aws.amazon.com/appconfig/latest/userguide/appconfig-creating-configuration-and-profile.html#appconfig-creating-configuration-and-profile-feature-flags) in future releases.
```
```suggestion
```

In this case, it's best to wait for customer demand via feature request (well known format). If we add that we're planning to add support now, we'll accidentally commit and prioritize over existing backlog items other customers asked.

heitorlessa commented on tests/e2e/parameters/infrastructure.py:
```diff
@@ -3,34 +3,31 @@
+        service_name = build_service_name()
+
+        cfn_application = appconfig.CfnApplication(
+            self.stack, id="appconfig-app", name=f"appe2e{service_name}", description="appe2e"
```
```suggestion
            self.stack, id="appconfig-app", name=f"powertools-e2e-{service_name}", description="Lambda Powertools End-to-End testing for AppConfig"
```

making it more explicit in the event of dangling resources caused by network split

heitorlessa commented on tests/e2e/parameters/infrastructure.py:
```diff
@@ -3,34 +3,31 @@
+            self.stack,
+            "appconfig-env",
+            application_id=cfn_application.ref,
+            name=f"enve2e{service_name}",
```
```suggestion
            name=f"powertools-e2e{service_name}",
```

heitorlessa commented on tests/e2e/parameters/infrastructure.py:
```diff
@@ -3,34 +3,31 @@
+            "appconfig-env",
+            application_id=cfn_application.ref,
+            name=f"enve2e{service_name}",
+            description="enve2e",
```
```suggestion
            description="Lambda Powertools End-to-End testing environment",
```

heitorlessa reviewed (approved):
THANK YOU! I'm glad we're completing this and we're ready for the grooming part before launching V2 now.

Last suggestions I'm committing myself is on being more explicit on AppConfig Config/Env Name and Description. If a network split happens, this will leave dangling resources making it difficult to trace back to what they are for (`appe2e` vs `powertools-e2e-x`).

heitorlessa committed:
docs(parameters): remove premature backlog commitment

heitorlessa committed:
chore: make resource naming traceable

heitorlessa committed:
chore: make resource naming traceable

heitorlessa committed:
chore: make resource naming traceable
